asf_heat_pump_suitability/pipeline/transform/outdoor_space.py

```diff
@@ -0,0 +1,190 @@
+    building_intersections_gdf: gpd.GeoDataFrame, land_parcels_gdf: gpd.GeoDataFrame
+) -> gpd.GeoDataFrame:
+    """
+    Generate GeoDataFrame of land extents with total and max contiguous outdoor space areas.
+
+    Args:
+        building_intersections_gdf (gpd.GeoDataFrame): intersections of land parcel polygons and building footprint polygons
+        land_parcels_gdf (gpd.GeoDataFrame): land parcel polygons
+
+    Returns:
+        gpd.GeoDataFrame: original land parcel polygons with calculated total and max contiguous outdoor space area (m2)
+    """
+    # Get land minus buildings - MultiPolygons will be created for land parcels which get split into multi-parts
+    land_minus_buildings = gpd.overlay(
+        land_parcels_gdf,
+        building_intersections_gdf,
+        how="difference",
+        keep_geom_type=False,
+    )
+
+    # Explode multi-part geometries into single part geometries to get largest piece
+    land_minus_buildings_parts = land_minus_buildings.explode(index_parts=False)
+    land_minus_buildings_parts["outdoor_space_area_m2"] = (
+        land_minus_buildings_parts.geometry.area
+    )
+
+    # Keep max size and total outdoor area
+    outdoor_space_df = (
+        land_minus_buildings_parts.groupby("NATIONALCADASTRALREFERENCE")
+        .agg(
+            max_contiguous_outdoor_space_area_m2=("outdoor_space_area_m2", max),
+            total_outdoor_space_area_m2=("outdoor_space_area_m2", sum),
+        )
+        .reset_index(
+            # Drop duplicates in case there are multiple max values
+        )
+        .drop_duplicates(subset="NATIONALCADASTRALREFERENCE")
+    )
+
+    # Merge full land extent geometry back on
+    outdoor_space_gdf = gpd.GeoDataFrame(
+        outdoor_space_df.merge(
+            land_parcels_gdf[["NATIONALCADASTRALREFERENCE", "geometry"]],
+            how="left",
+            on="NATIONALCADASTRALREFERENCE",
+        )
+    )
+
+    return outdoor_space_gdf
+
+
+def deduplicate_df_outdoor_space(df: pl.DataFrame) -> pl.DataFrame:
+    """
+    Deduplicate UPRNs matched to multiple land extents by keeping the one with the smallest total area.
+
+    Args:
+        df (pl.DataFrame): UPRNs with outdoor space estimates
+
+    Returns:
+        pl.DataFrame: deduplicated UPRNs with outdoor space estimates
+    """
+    df = df.with_columns(pl.col("UPRN").is_duplicated().alias("UPRN_duplicated"))
+
+    _deduplicated_df = (
+        df.filter(pl.col("UPRN_duplicated"))
+        .with_columns(
+            min_total=pl.col("total_outdoor_space_area_m2")
+            .min()
+            .over("NATIONALCADASTRALREFERENCE")
+        )
+        .filter(
+            # Get smallest outdoor space
+            pl.col("total_outdoor_space_area_m2")
+            == pl.col("min_total")
+            # Deduplicate - any duplicates will now (most likely) have the same total outdoor space
+        )
+        .unique(subset="NATIONALCADASTRALREFERENCE")
+        .drop("min_total")
+    )
```

Comment:
```suggestion
    _deduplicated_df = (
        df.filter(pl.col("UPRN_duplicated"))
        .with_columns(
            min_total=pl.col("total_outdoor_space_area_m2")
            .min()
            .over("UPRN")
        )
        .filter(
            # Get smallest outdoor space
            pl.col("total_outdoor_space_area_m2")
            == pl.col("min_total")
            # Deduplicate - any duplicates will now (most likely) have the same total outdoor space
        )
        .unique(subset="UPRN")
        .drop("min_total")
    )
```

As we've discussed offline, this processing should be over the `UPRN` column rather than over `NATIONALCADASTRALREFERENCE`.